Add support for a default file extension for view names

dmaidaniuk opened the pull request:
See an issue https://github.com/mvc-spec/ozark/issues/151 for details.

dmaidaniuk commented:
@chkal,
Travis build for this PR failed due to reached timeout for `glassfish-module` job which took more than 49 min to execute. All this time was spent to download artifacts from Maven repository. Do you know why this job executed so slowly? Other jobs in this build executed much faster.

chkal commented:
@dmaidaniuk You can ignore these timeouts for now. That's a Travis issue. I'm already in contact with the Travis team and they are working on a fix.

chkal commented on core/src/main/java/org/mvcspec/ozark/core/ViewResponseFilter.java:
```diff
@@ -80,6 +81,7 @@
 public class ViewResponseFilter implements ContainerResponseFilter {
 
     private static final String REDIRECT = "redirect:";
+    private static final String DOT_CHAR = ".";
```
Not sure how others feel about that, but defining a constant for the dot character isn't really required. I would be fine with inlining it.

dmaidaniuk replied:
Fixed.

chkal commented on core/src/main/java/org/mvcspec/ozark/core/ViewResponseFilter.java:
```diff
@@ -111,6 +116,7 @@ public void filter(ContainerRequestContext requestContext,
 
         final Method method = resourceInfo.getResourceMethod();
         final Class<?> returnType = method.getReturnType();
+        final String defaultExtension = ozarkConfig.getDefaultViewFileExtension();
```
Perhaps we should let `appendExtensionIfRequired` lookup the default extension instead of doing this here? This way we won't need to pass this into every call to `appendExtensionIfRequired`. I think that's fine because effectively `appendExtensionIfRequired` will only be called one per request so there won't be any performance befit of reading the value here.

The only argument for doing the lookup of the default value outside of `appendExtensionIfRequired` is that it simplifies unit testing. But we could work around this by doing something like this:

```java
// call this method in the filter
private static String appendExtensionIfRequired(String viewName) {
  return appendExtensionIfRequired(viewName, ozarkConfig.getDefaultViewFileExtension());
}

protected static String appendExtensionIfRequired(String viewName, String defaultExtension) {
  // same as in your pull request
}
```

dmaidaniuk replied:
Done.

chkal commented on core/src/main/java/org/mvcspec/ozark/core/ViewResponseFilter.java:
```diff
@@ -173,6 +179,24 @@ public void filter(ContainerRequestContext requestContext,
+        if (!viewName.contains(DOT_CHAR) && !viewName.contains(":")) {
+            resultView += DOT_CHAR + defaultExtension;
+        } else if (viewName.endsWith(DOT_CHAR)) {
+            throw new IllegalArgumentException("Invalid format for view name: " + viewName);
```
I don't think that we should fail here. Of course view names with trailing dots are weird, but technically they will work. So why not removing this else block?

dmaidaniuk replied:
Else block removed.

chkal commented on core/src/main/java/org/mvcspec/ozark/core/ViewResponseFilter.java:
```diff
@@ -173,6 +179,24 @@ public void filter(ContainerRequestContext requestContext,
+        }
+
+        String resultView = viewName;
+        if (!viewName.contains(DOT_CHAR) && !viewName.contains(":")) {
```
Could you explain the `!viewName.contains(":")` here? Not sure if this is really needed. Especially because I know at least one view engine which supports view names like `react:some-folder/foobar.jsp`.

dmaidaniuk replied:
Exactly for view engines like mentioned I skipped the default extension attaching in case if the character `:` available in the incoming view name. So, view names like `react:some-folder/foobar.jsp` or `react:some-folder/foobar` will be excluded from handling and remain unchanged.
Do you think that it is an incorrect assumption and we still need to process such view names?

chkal replied:
I don't see a reason why we should not apply the default file extension mechanism if the view name contains `:`. Actually a view name like `react:foobar.jsp` would mean something like "use the react view engine but render the view using a default JSP view engine". So why not apply the default mechanism in this case?

Of course `redirect:` needs special treatment, but this is already done in the first `if` block. So I think it is safe to add the default extension if there is no `.` char in the view name. So I would prefer to simply remove the `!viewName.contains(":")` part.

dmaidaniuk replied:
OK, I got it. I made the required changes.

chkal commented:
Oh, I forgot to mention one thing. I think I was able to fix the Travis errors you were seeing in #207. But there is no real need to rebase to master just to get this fix. I can restart the failed Travis builds manually if required and usually this fixes the problem with the build. But rebasing world of course also work. :-)

dmaidaniuk commented:
Hi @chkal,

I fixed almost all issues, which you mentioned in review. Only one left here: https://github.com/mvc-spec/ozark/pull/204#discussion_r219689624
I added my explenation for it. Please take a look.
Also, build on Travis passed now after your fix.

Regards,
Dmytro

chkal commented:
Awesome! Thank you very much for applying the changes. I added a comment to the last remaining one.

dmaidaniuk commented:
@chkal I made the last requested change. Build on Travis passed.

chkal commented:
Awesome! Thanks @dmaidaniuk for all your help with this. I'll merge this now! :beers:

dmaidaniuk commented:
@chkal you're welcome. Glad to help you with this task. I guess that original issue #151 could be also closed now.

Daniel-Dos commented:
Very good. 

thanks @dmaidaniuk  and @chkal  . 🍻
